### ISSUE TITLE
Proper composer autoloading

### ISSUE THREAD

**Issue by slaFFik** [woocommerce/action-scheduler#464]
I decided to merge several issues in one and list them all here and provide a solution.
I'm writing here from point of view of a developer, who already uses composer in their own plugin and wants to include AS as yet another package.

Right now AS doesn't properly use composer autoload. Perhaps, there are some reasons for that, I don't know them.
Because of the way AS 3.1.0 is loading ([using classmap](https://github.com/woocommerce/action-scheduler/blob/400725e8f80086f9f584a17b17b5352e27df5b52/composer.json#L37), implemented in #434, + manual file require), it's not possible to use a correct way to install packages and use the ordinary composer autoload functionality.

So, this doesn't work:
```php
require_once __DIT__ . 'vendor/autoload.php';
```
You need to explicitly require AS main file, and do it BEFORE composer autoload, like this:
```php
require_once __DIR__ . '/vendor/woocommerce/action-scheduler/action-scheduler.php';
require_once __DIR__ . '/vendor/autoload.php';
```
This is only relevant for AS 3.1.0 (because of #434).
AS 3.0.1 works fine when included via composer and initiated directly like this:
```php
require_once __DIR__ . '/vendor/autoload.php';
require_once __DIR__ . '/vendor/woocommerce/action-scheduler/action-scheduler.php';
```

With this v3.1.0 approach, composer is not used at all to load AS, everything is done in AS and its `classmap` is useless, afaik.

My investigation in #463 gave me that extra information, and understanding, that perhaps this situation can be improved because I'm not the only one (see #445 as an example).

So, I suggest to remove `"classmap"` and add `"files"`:
```
"autoload": {
    "files": ["action-scheduler.php"]
},
```

This way a proper way to include packages, using
```php
require_once __DIT__ . 'vendor/autoload.php';
```
will be possible, without the need to require `action-scheduler.php` in own plugin.

**Comment by rrennick** [comment]
@slaFFik Thanks for the explanation. The way Action Scheduler currently works when there are multiple plugins active that include AS is

- `aa` has AS `2.2.5`
- `bb` has AS `3.1.0`
- `cc` has AS `3.0.1`

When WP loads each plugin (in alphabetical order) 
- `action-scheduler.php` registers its version with a callback for initializing that version
- on `plugins_loaded` AS runs the callback of the latest version
- the callback registers the autoloader for that version

In this example, only the autoloader of `bb` would be registered. 

If AS were updated to work the way you have requested then the autoloaders for all 3 versions would be registered. If all 3 plugins stuck strictly to using `as_*` functions then they would all be using AS 2.2.5 because it is the first registered autoloader. If either `bb` or `cc` use `as_enqueue_async_action()` it would trigger fatal errors as it calls functions that do not exist in the 2.2.5 classes.

**Comment by XedinUnknown** [comment]
**TL;DR**
This hacky way of loading the library is making normal usage of it much more complicated. Please, is there a way to go around this? If not, could it be made optional?

**Details**
Which version of a package to use is not the concern of that package itself. It is the concern of the consumer of that package. I see 2 main use-cases for how AS can be used:

1. As part of WooCommerce itself. WC seems to have a UI that integrates with AS, and many people use AS in conjunction with WC. So, just ship the latest version of AS with your latest version of WC, and any plugin that integrates with WC will automatically have AS to use.
2. As a separate package. When my PluginA uses Guzzle, and PluginB uses Guzzle too, nobody thinks that Guzzle should, instead of registering its autoload configuration as usual, do magic with latest version selection, etc. That's because the developers of Guzzle realize that which version of their package the consumer should use is _not for them to decide_. If both of the above-mentioned plugins use Guzzle, then they should do something about the possible collision, such as [scoping].

The kind of hacks with hooks that are being used right now are preventing normal usage of this great library, because it requires additional configuration and considerations. Why should I have to use _any_ hook to use the schedule, even more so - something as late as `init`? This adds unnecessary complexity to the consuming software.

[scoping]: https://inpsyde.com/en/package-management-in-wordpress-introduction-solutions/#custom_html-3:~:text=Automatic%20Refactoring%20During%20Build

**Comment by rrennick** [comment]
> is there a way to go around this? If not, could it be made optional?

@XedinUnknown No, there isn't a way around it. I explained the reason in detail in #471. Action Scheduler is included as a library in dozens of plugins. An install can only have one instance of AS functionally active. Redeclaring functions or classes triggers fatal errors.

> WC seems to have a UI that integrates with AS

That UI is in Action Scheduler.

> just ship the latest version of AS with your latest version of WC

The base assumption this relies on is that all installs update their plugins as soon as there is an update available.

> Why should I have to use any hook to use the schedule, even more so - something as late as init?

The original data store for AS was in the posts tables. You cannot register post types in WordPress until `init`.

> use different incompatible versions of AS

In the latest version of AS, all of the `as_*` function signatures are compatible with every previous version. IOW, if a plugin was written to work with AS 2.0.0, it works with 3.1.6.

**Comment by XedinUnknown** [comment]
> An install can only have one instance of AS functionally active. Redeclaring functions or classes triggers fatal errors

Shouldn't be a re-declaration if relying on autoloading.

> The base assumption this relies on is that all installs update their plugins as soon as there is an update available.

How so? IMHO, it only implies that, from a certain point in time, everyone will have AS 3.0+.

> The original data store for AS was in the posts tables. You cannot register post types in WordPress until init.

Does this mean that with AS 3.0+ it's possible to use the scheduler earlier, like on `plugins_loaded`?

> IOW, if a plugin was written to work with AS 2.0.0, it works with 3.1.6

Ok, but the API can break in the future too.
